### PR TITLE
feat(extensions): add helper header above extensions search

### LINKED ISSUE #2132: Add help information to the extension search view
Show a yellow or red call to action in the extension search view informing the user of all alternative extension installation methods.

### PR DESCRIPTION
**Pull request by jsjoeio**
Add a short message above the search box on the Extensions panel. This helps explain the extension divergence to the user.

If they click dismiss, it stores an item in localStorage to prevent the message from showing up on subsequent loads.

Builds off the work previously done by @cmoog in #2185 

## Screenshots

And it uses the correct colors based on theme:

https://user-images.githubusercontent.com/3806031/104351495-b3b05b00-54c2-11eb-8c4a-fb0a4d90ac7e.MP4

## Checklist
- [x] tested locally
- [x] double-check link styles (see Welcome file for example)
- [x] squash long GitHub commits
Fixes #2132

### CODE REVIEW

File: doc/ipad.md
Comment (nhooyr) [inline doc/ipad.md:10]: I believe this PR needs to be rebased on master.

I fixed this in https://github.com/cdr/code-server/commit/4185675338767c4b73eca7a23e80488bece364e6

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (nhooyr) [inline lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:406]: I can't comment on this in github but the commit message here should be hard wrapped to 80 lines.
It's a common standard when using git.

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (nhooyr) [inline lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:409]: was this intended?

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (code-asher) [inline lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: Why do we need the timeout?

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (code-asher) [inline lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:428]:
```suggestion
			These extensions are not official. Find additional open-source extensions
```

I feel like this might make it a bit more clear it's something separate from the list you're seeing in the editor.

**Review by code-asher — APPROVED** [review]
Looks great!

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (code-asher) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: Oh also we could tweak this a bit so it doesn't depend on `this.root`. Probably fine either way but since `this.root` could be reassigned I guess it's technically safer this way:

```suggestion
				setTimeout(() => helperHeader.remove());
```

**Review by nhooyr — CHANGES_REQUESTED** [review]

**Comment by nhooyr** [comment]
Love the gif btw 🔥

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: > Why do we need the timeout?

Good question...I am not sure why I added that 🤔 I'll remove it.

Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: > Oh also we could tweak this a bit so it doesn't depend on `this.root`. Probably fine either way but since `this.root` could be reassigned I guess it's technically safer this way:
> 
> 
> 
> ```suggestion
> 
> 				setTimeout(() => helperHeader.remove());
> 
> ```

Great suggestion!

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:406]: Ah...tbh I was trying to write more descriptive git commit messages after looking at some of yours. Didn't realize I had to wrap it myself. Is there a setting I can change in my git configuration to prevent this from happening in the future?

Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:406]: I wrapped it using [this method](https://stackoverflow.com/a/3033455/3015595) but it would be great to configure that automatically for next time.

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:409]: I think I was just adding space to make it more readable, but I'll clean it up

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (nhooyr) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts]: Should be able to configure VS Code to wrap for you. In fact it should be the default. Are you committing via vim on the CLI or VS Code? For vim there's some settings you have to set to make it automatic but I don't remember off the top of my head.

**Comment by jsjoeio** [comment]
> rsync: change_dir "/github/workspace//typings/pluginapi.d.tsrelease" failed: No such file or directory (2)
2057
> rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1207) [sender=3.1.3]

I’m not totally sure why the `release` step is failing. Any ideas?

**Comment by code-asher** [comment]
Ahh, sorry about the failed release step, it's my fault. I pushed a fix to master, please feel free to rebase

File: lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts
Comment (jsjoeio) [reply lib/vscode/src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts]: For this specific commit, it was via Vim in the VSCode terminal. No worries at all. Thanks!

**Comment by jsjoeio** [comment]
Bump @nhooyr - ready for a second review

**Comment by nhooyr** [comment]
Can confirm things work for me!

But the links appear to be styled very hard to read for some reason?

Are we maybe missing a class on the links or some styles?

Screen Shot 2021-01-11 at 12 43 38 PM

**Comment by nhooyr** [comment]
Like there's little contrast compared to the other links like: 
Screen Shot 2021-01-11 at 12 44 42 PM

**Comment by nhooyr** [comment]
@jsjoeio make sure to squash the really long Update file commits generated by github.

**Comment by jsjoeio** [comment]
> Can confirm things work for me!
> 
> But the links appear to be styled very hard to read for some reason?
> 
> Are we maybe missing a class on the links or some styles?

Hmm! Good catch. Let me look into it.

**Comment by jsjoeio** [comment]
> @jsjoeio make sure to squash the really long Update file commits generated by github.

You got it!

**Comment by jsjoeio** [comment]
> Like there's little contrast compared to the other links like:

I’m having trouble figuring this out on iPad since I don’t have access to DevTools to inspect the styles 🤔 I’m looking at the Welcome page and don’t see anything jumping out at me. The links in the list have no classes so i’m not sure where the link color is coming from.

4FDD2F99-E546-4194-8536-85CDC6745F09

The easiest solution would be to hard-code the link style to match the rest of code-server but I worry that wouldn’t work well across various themes. I’ll keep investigating.

**Comment by jsjoeio** [comment]
Well...hard-coding it works. Not sure what else to try. Any thoughts/suggestions @nhooyr ?

8B5C8CEE-B40C-446B-A4CD-271AE6B82FF5

**Review by nhooyr — COMMENTED** [review]
Happy to merge as is but one downside of hard coding the link color like we have is that the `:visited` style doesn't take any affect and so the links don't turn purple anymore which sucks.

I understand @code-asher has some ideas to fix this?

**Comment by nhooyr** [comment]
Or one more I should say, like you said themes will also be a problem. So we can't merge as is then.

**Comment by jsjoeio** [comment]
Yeah exactly. I'll move this back to draft state then re-request a review when it's ready!

**Comment by jsjoeio** [comment]
Fixed the color issues (H/T to @code-asher for pointing me in the right direction)

BEFORE
D01570C0-87A7-47D6-8CC8-7017963CEBA4

AFTER
E59F1722-7B11-4F6B-9A1A-F4518C55B4AE

**Review by nhooyr — APPROVED** [review]
perfecto!! 🔥 

I noticed that `:visited` isn't supported by any of VS Code's other UI links so it's fine that we don't here too.

**Comment by jsjoeio** [comment]
Oh didn't even think about that - thanks for checking! Let's get this bad boy merged in!

**Comment by jsjoeio** [comment]
After i fix the lint issues :)

**Pull request merged**
